File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    List<List<Integer>> ans;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        ans=new ArrayList<>();
        solve(candidates,target,0,0,new ArrayList<>());
        return ans;
    }
    void  solve(int []arr,int k,int sum,int i,List<Integer> him){
        if(sum==k){
            ans.add(new ArrayList<>(him));
            return;
        }
        if(i>=arr.length){
            return ;
        }
        if(k-sum>=arr[i])
      {  him.add(arr[i]);
        solve(arr,k,sum+arr[i],i,him);
        him.remove(him.size()-1);}
        solve(arr,k,sum,i+1,him);
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def __init__(self):
        self.a=[]
        self.b=[]
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.solve(candidates,0,target,0)
        return self.a
    def solve(self,arr,i,target,sum):
        if (i==len(arr) or sum>target):
            return
        if(sum==target):
            self.a.append(self.b.copy())
            return
        self.b.append(arr[i])
        self.solve(arr,i,target,sum+arr[i])
        self.b.pop()
        self.solve(arr,i+1,target,sum)




        